### PR TITLE
8342607: Enhance register printing on x86_64 platforms

### PR DESCRIPTION
**Pull request by TheRealMDoerr**
There are some situations in which the XMM registers are relevant to understand errors. E.g. C2 compiler uses them to spill GPR values (see https://github.com/openjdk/jdk/blob/d6eddcdaf92f2352266ba519608879141997cd63/src/hotspot/cpu/x86/x86_64.ad#L1312), so they may contain Oops etc. We may consider searching and printing the content for Oops in a future RFE.
I've implemented [JDK-8342607](https://bugs.openjdk.org/browse/JDK-8342607) such that linux and Windows show the same output format. (Skipped Intel-Mac because Apple has stopped shipping that platform. I don't have it and I'm not familiar with it.)

Example output (linux):
```
Registers:
RAX=0x00007fea8bdb3000, RBX=0x00007fea8b48d5d4, RCX=0x00007fea8b4d2255, RDX=0x0000000000000340
RSP=0x00007fea897d0b60, RBP=0x00007fea897d0b90, RSI=0x00007fea8b5f1448, RDI=0x00000000e0000000
R8 =0x00007fea8b48d5d4, R9 =0x0000000000000006, R10=0x00007fea8bb4b500, R11=0x00007fea7cc2f120
R12=0x0000000000000000, R13=0x00007fea897d0bc0, R14=0x00007fea897d0c50, R15=0x00007fea8402c9c0
RIP=0x00007fea8ac008e5, EFLAGS=0x0000000000010246, CSGSFS=0x002b000000000033, ERR=0x0000000000000006
  TRAPNO=0x000000000000000e

XMM[0]=0x0000000000000000 0x0000000000000000
XMM[1]=0x00007fea3c034200 0x0000000000000000
XMM[2]=0x00000000fffffffe 0x00007fea8402c9c0
XMM[3]=0x00007fea7c3d6608 0x0000000000000000
XMM[4]=0x00007f0000000000 0x0000000000000000
XMM[5]=0x00007fea897d0fe8 0x00007feaffffffff
XMM[6]=0x0000000000000000 0x00007fea897d0f98
XMM[7]=0x0202020202020202 0x0000000000000000
XMM[8]=0x0000000000000000 0x0202020202020202
XMM[9]=0x666e69206e6f6974 0x0000000000000000
XMM[10]=0x0000000000000000 0x6e6f6974616d726f
XMM[11]=0x0000000000000001 0x0000000000000000
XMM[12]=0x00007fea8b684400 0x0000000000000001
XMM[13]=0x0000000000000000 0x0000000000000000
XMM[14]=0x0000000000000000 0x0000000000000000
XMM[15]=0x0000000000000000 0x0000000000000000
  MXCSR=0x0000037f
```

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8342607](https://bugs.openjdk.org/browse/JDK-8342607): Enhance register printing on x86_64 platforms (**Enhancement** - P4)

### Reviewers
 * [Matthias Baesken](https://openjdk.org/census#mbaesken) (@MBaesken - **Reviewer**) Review applies to [b3ec6143](https://git.openjdk.org/jdk/pull/21615/files/b3ec6143878d5d4c9a005264e3b4b8baabfbe040)
 * [Richard Reingruber](https://openjdk.org/census#rrich) (@reinrich - **Reviewer**)
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**)

### Contributors
 * Richard Reingruber `<rrich@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21615/head:pull/21615` \
`$ git checkout pull/21615`

Update a local copy of the PR: \
`$ git checkout pull/21615` \
`$ git pull https://git.openjdk.org/jdk.git pull/21615/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21615`

View PR using the GUI difftool: \
`$ git pr show -t 21615`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21615.diff">https://git.openjdk.org/jdk/pull/21615.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21615#issuecomment-2426876074)

### CODE REVIEW

File: src/hotspot/os_cpu/windows_x86/os_windows_x86.cpp
Comment (TheShermanTanker) [inline src/hotspot/os_cpu/windows_x86/os_windows_x86.cpp:444]: Using a more specific cast is generally better, in this case since it's a pointer to pointer, reinterpret\_cast might suffice. And if that is used, the const is not required (I think)

Comment (TheShermanTanker) [reply src/hotspot/os_cpu/windows_x86/os_windows_x86.cpp:444]: Actually, scratch that. The const is probably needed even with reinterpret\_cast. I got it mixed up with something else

**Comment by MBaesken** [comment]
Looks good to me, but I could live well without that indentation of the MXCSR value.

**Comment by reinrich** [comment]
Hi Martin,
is there documentation on which you base the implementation? My quick search wasn't very successful.
How do we know that the printed values are correct?

**Comment by TheRealMDoerr** [comment]
> Hi Martin, is there documentation on which you base the implementation? My quick search wasn't very successful. How do we know that the printed values are correct?

You can find the data structures here: https://github.com/bminor/glibc/blob/dcad78507433a9a64b8b548b19e110933f8d939a/sysdeps/x86_64/sys/ucontext.h#L106

**Comment by MBaesken** [comment]
MS documents also the CONTEXT structure with various register values, including XMM
https://learn.microsoft.com/de-de/windows/win32/api/winnt/ns-winnt-context

**Comment by reinrich** [comment]
Well, with my quick-search I found information like that too. I don't see though, how it helps understand that the printed values are correct.

**Review by MBaesken — APPROVED** [review]

**Comment by TheRealMDoerr** [comment]
> Well, with my quick-search I found information like that too. I don't see though, how it helps understand that the printed values are correct.

I think the only way to verify that the printed values are correct is testing it. I'm currently experimenting with setting up XMM values and then crashing intentionally. I can see that my values get printed, but on linux, XMM[0] seems to be always 0 and the other registers off by one. Strange.

**Comment by reinrich** [comment]
Thanks Martin for doing this. I actually wanted to attempt writing a test myself before I read your message.

> I can see that my values get printed, but on linux, XMM[0] seems to be always 0 and the other registers off by one. Strange.

Strange indeed.

**Comment by tstuefe** [comment]
> > Well, with my quick-search I found information like that too. I don't see though, how it helps understand that the printed values are correct.
> 
> I think the only way to verify that the printed values are correct is testing it. I'm currently experimenting with setting up XMM values and then crashing intentionally. I can see that my values get printed, but on linux, XMM[0] seems to be always 0 and the other registers off by one. Strange.

I got curious, then confused, and looked. See https://stackoverflow.com/questions/43415882/reading-sse-registers-xmm-ymm-in-a-signal-handler - even though it refers to ymm registers. The answer seems to indicate that the glibc structure is not what you should be using, but going via uc_mcontext seems to be the right way.

**Comment by TheRealMDoerr** [comment]
Thanks for your input @reinrich, @tstuefe. I've tried taking the values from `uc->uc_mcontext.fpregs->_xmm[i]`. This seems to work fine if we got a signal. However, "should_not_reach_here()" and friends don't trigger any signal. They call `MacroAssembler::debug64`. Surprisingly, my old code based on `uc->__fpregs_mem._xmm[i].element[0]` works for them (besides the off-by-one issue), but the new proposal doesn't. Any idea?

**Comment by reinrich** [comment]
For the record: the kernel sets the `fpregs` pointer [here](https://github.com/torvalds/linux/blob/eb26cbb1a754ccde5d4d74527dad5ba051808fad/arch/x86/kernel/signal_64.c#L128) :)

**Comment by tstuefe** [comment]
> Thanks for your input @reinrich, @tstuefe. I've tried taking the values from `uc->uc_mcontext.fpregs->_xmm[i]`. This seems to work fine if we got a signal. However, "should_not_reach_here()" and friends don't trigger any signal. They call `MacroAssembler::debug64`. 

odd. debug64 ends up calling fatal, which should use the assertion poison page to get a valid context.

**Comment by reinrich** [comment]
> > Thanks for your input @reinrich, @tstuefe. I've tried taking the values from `uc->uc_mcontext.fpregs->_xmm[i]`. This seems to work fine if we got a signal. However, "should_not_reach_here()" and friends don't trigger any signal. They call `MacroAssembler::debug64`.
> 
> odd. debug64 ends up calling fatal, which should use the assertion poison page to get a valid context. see debug.hpp.

The [`ucontext_t` is memcpyed](https://github.com/openjdk/jdk/blob/f0b130e54f33d3190640ce33c991e35f27e9f812/src/hotspot/share/utilities/debug.cpp#L727). The `fpregs` pointer in the copy is probably invalid when used.

**Comment by tstuefe** [comment]
> > Thanks for your input @reinrich, @tstuefe. I've tried taking the values from `uc->uc_mcontext.fpregs->_xmm[i]`. This seems to work fine if we got a signal. However, "should_not_reach_here()" and friends don't trigger any signal. They call `MacroAssembler::debug64`.
> 
> odd. debug64 ends up calling fatal, which should use the assertion poison page to get a valid context. see debug.hpp.

to extend: we trigger an artificial segfault in fatal with the poison page, and grab that context. Generally that works. Question is why not here.

**Comment by tstuefe** [comment]
> > > Thanks for your input @reinrich, @tstuefe. I've tried taking the values from `uc->uc_mcontext.fpregs->_xmm[i]`. This seems to work fine if we got a signal. However, "should_not_reach_here()" and friends don't trigger any signal. They call `MacroAssembler::debug64`.
> > 
> > 
> > odd. debug64 ends up calling fatal, which should use the assertion poison page to get a valid context. see debug.hpp.
> 
> The [`ucontext_t` is memcpyed](https://github.com/openjdk/jdk/blob/f0b130e54f33d3190640ce33c991e35f27e9f812/src/hotspot/share/utilities/debug.cpp#L727). The `fpregs` pointer in the copy is probably invalid when used.

oh good point! I remember dealing with a similar problem on PPC years ago and implementing a deep copy somehow.

**Comment by reinrich** [comment]
The context is copied in the signal handler and used [here](https://github.com/openjdk/jdk/blob/f0b130e54f33d3190640ce33c991e35f27e9f812/src/hotspot/share/utilities/debug.cpp#L208) after the signal handler has returned. The copied `fpregs` pointer referes still to original which is/was located in the signal handler caller frame. You need to set it to the `__fpregs_mem` in the copy.

**Comment by reinrich** [comment]
> The context is copied in the signal handler and used [here](https://github.com/openjdk/jdk/blob/f0b130e54f33d3190640ce33c991e35f27e9f812/src/hotspot/share/utilities/debug.cpp#L208) after the signal handler has returned. The copied `fpregs` pointer referes still to original which is/was located in the signal handler caller frame. You need to set it to the `__fpregs_mem` in the copy.

Correction: the kernel has aligned `fpregs` so it doesn't point precisely to `__fpregs_mem` (that's why Martins version isn't working). The copied `fpregs` must point to the same offset in the `ucontext_t` copy as the original `fpregs` pointer in the original `ucontext_t`.

**Comment by TheRealMDoerr** [comment]
Thanks for figuring this out! So, there's a bit more to repair :-) I'll take a look.

**Comment by TheRealMDoerr** [comment]
It works when I repair the copied `uc_mcontext` (see 2nd commit). Please take a look. Note that I'm using little endian style: print the high order half first which is at offset 8, then the low order half (offset 0).
We should also backport this fix.

**Comment by TheRealMDoerr** [comment]
Also note that the copy fix above is kind of a hack. Should we use something like the following?
```diff
diff --git a/src/hotspot/share/utilities/debug.cpp b/src/hotspot/share/utilities/debug.cpp
index 988e5dddd90..1389654b4b6 100644
--- a/src/hotspot/share/utilities/debug.cpp
+++ b/src/hotspot/share/utilities/debug.cpp
@@ -731,7 +731,9 @@ static void store_context(const void* context) {
 #if defined(PPC64)
   *((void**) &g_stored_assertion_context.uc_mcontext.regs) = &(g_stored_assertion_context.uc_mcontext.gp_regs);
 #elif defined(AMD64)
-  *((void**) &g_stored_assertion_context.uc_mcontext.fpregs) = &(g_stored_assertion_context.uc_mcontext.fpregs);
+  // In the copied version, fpregs should point to the copied contents. Preserve the offset.
+  intptr_t offset = (address)(void*)(g_stored_assertion_context.uc_mcontext.fpregs) - (address)context;
+  *((void**) &g_stored_assertion_context.uc_mcontext.fpregs) = (void*)((address)(void*)&g_stored_assertion_context + offset);
 #endif
 #endif
 }
```

**Comment by reinrich** [comment]
> Also note that the copy fix above is kind of a hack. Should we use something like the following?
> 
> ```diff
> diff --git a/src/hotspot/share/utilities/debug.cpp b/src/hotspot/share/utilities/debug.cpp
> index 988e5dddd90..1389654b4b6 100644
> --- a/src/hotspot/share/utilities/debug.cpp
> +++ b/src/hotspot/share/utilities/debug.cpp
> @@ -731,7 +731,9 @@ static void store_context(const void* context) {
>  #if defined(PPC64)
>    *((void**) &g_stored_assertion_context.uc_mcontext.regs) = &(g_stored_assertion_context.uc_mcontext.gp_regs);
>  #elif defined(AMD64)
> -  *((void**) &g_stored_assertion_context.uc_mcontext.fpregs) = &(g_stored_assertion_context.uc_mcontext.fpregs);
> +  // In the copied version, fpregs should point to the copied contents. Preserve the offset.
> +  intptr_t offset = (address)(void*)(g_stored_assertion_context.uc_mcontext.fpregs) - (address)context;
> +  *((void**) &g_stored_assertion_context.uc_mcontext.fpregs) = (void*)((address)(void*)&g_stored_assertion_context + offset);
>  #endif
>  #endif
>  }
> ```

Looks like what I ment in [my comment above](https://github.com/openjdk/jdk/pull/21615#issuecomment-2434816007). I was wondering why you ignored it.

File: src/hotspot/share/utilities/debug.cpp
Comment (reinrich) [inline src/hotspot/share/utilities/debug.cpp:735]: Reads better and is safer:
```suggestion
  size_t fpregs_offset = pointer_delta(((const ucontext_t*)context)->uc_mcontext.fpregs, context, 1);
```

File: src/hotspot/os_cpu/linux_x86/os_linux_x86.cpp
Comment (reinrich) [inline src/hotspot/os_cpu/linux_x86/os_linux_x86.cpp:551]: Better check if `fpregs` points into `*uc` before dereferencing it (asserting that it actually does).

**Review by reinrich — COMMENTED** [review]
I've done a little bit of adhoc testing. Seems to work now. We could add a new test to test/hotspot/jtreg/runtime/ErrorHandling but on the other hand I don't think it is really necessary.
Cheers, Richard.

File: src/hotspot/os_cpu/linux_x86/os_linux_x86.cpp
Comment (TheRealMDoerr) [reply src/hotspot/os_cpu/linux_x86/os_linux_x86.cpp]: An assertion is not ideal. We don't want to run into it while printing an hs_err file. I've added a sanity check and print another message instead if it's bad.

**Comment by TheRealMDoerr** [comment]
I only don't like that pointer_delta contains an assertion.

**Comment by reinrich** [comment]
> I only don't like that pointer_delta contains an assertion.

Hm, the code gets bloated if you try to avoid it. Maybe it's acceptable(?).
Avoiding it could look like this: https://github.com/reinrich/jdk/commit/4097a7fdccd0e393d5e9e2bdb6b3725e5bed4ae6

**Comment by TheRealMDoerr** [comment]
Maybe you're right and the assertion is acceptable. We should never get a broken context. And if it ever happens we may live with the assertion. Product builds should be fine.

File: src/hotspot/os_cpu/linux_x86/os_linux_x86.cpp
Comment (reinrich) [inline src/hotspot/os_cpu/linux_x86/os_linux_x86.cpp:547]: Isn't it compiling on MUSL? `ucontext_t` and `mcontext_t` seem to be equal looking at https://git.musl-libc.org/cgit/musl/tree/arch/x86_64/bits/signal.h
Also the context comes from the kernel and should have the same layout I thought.
If it isn't working then I guess the changes in debug.cpp won't work for MUSL either ;)

Comment (TheRealMDoerr) [reply src/hotspot/os_cpu/linux_x86/os_linux_x86.cpp]: Let's give it a try. See Commit nr. 6.

**Review by reinrich — APPROVED** [review]
Looks good,
Richard.

File: src/hotspot/os_cpu/linux_x86/os_linux_x86.cpp
Comment (tstuefe) [inline src/hotspot/os_cpu/linux_x86/os_linux_x86.cpp:550]: Could the register substructure live outside uc on x64? If so, it may be safer to 
```suggestion
  size_t fpregs_offset = (uc->uc_mcontext.fpregs >= uc) ? pointer_delta(uc->uc_mcontext.fpregs, uc, 1) : 0;
```
or similar, since the register substructure may precede uc which would make pointer_delta assert.

File: src/hotspot/share/utilities/debug.cpp
Comment (tstuefe) [inline src/hotspot/share/utilities/debug.cpp:735]: Interesting note, above I see we do similar adjustments for ppc (https://github.com/openjdk/jdk/commit/3e603a776ea9c5642cb0ec6b9105c6ff34b8f2b1)

**Review by tstuefe — APPROVED** [review]
Small nit. But thank you for fixing the assertion poison mechanism too.
I'll approve now and leave it up to you if you take my suggestion.

File: src/hotspot/os_cpu/linux_x86/os_linux_x86.cpp
Comment (TheRealMDoerr) [reply src/hotspot/os_cpu/linux_x86/os_linux_x86.cpp:550]: I think using 0 would require more changes to avoid accessing uc+0 which would be wrong. Richard and I already discussed about this above. We think it's acceptable. Some other projects claim that FP register substructure is inside the uc: https://github.com/mono/mono/blob/0f53e9e151d92944cacab3e24ac359410c606df6/mono/utils/mono-sigcontext.h#L263
We could also check the kernel code which writes it.

**Comment by TheRealMDoerr** [comment]
Thanks for the reviews!

File: src/hotspot/os_cpu/linux_x86/os_linux_x86.cpp
Comment (TheRealMDoerr) [reply src/hotspot/os_cpu/linux_x86/os_linux_x86.cpp]: After having read this, I have made the change. See commit nr. 7.

**Review by tstuefe — APPROVED** [review]
still good

File: src/hotspot/os_cpu/linux_x86/os_linux_x86.cpp
Comment (reinrich) [inline src/hotspot/os_cpu/linux_x86/os_linux_x86.cpp:552]: Why protect against the assertion in `pointer_delta` here and not also in `store_context`?
You could do the following in both methods:
```suggestion
  size_t fpregs_offset = ((address)uc->uc_mcontext.fpregs >= (address)uc) ?
                         pointer_delta(uc->uc_mcontext.fpregs, uc, 1) : sizeof(ucontext_t);
  if (fpregs_offset >= sizeof(ucontext_t)) {
```

Comment (TheRealMDoerr) [reply src/hotspot/os_cpu/linux_x86/os_linux_x86.cpp]: I've changed it such that we don't touch in this case.

File: src/hotspot/share/utilities/debug.cpp
Comment (reinrich) [inline src/hotspot/share/utilities/debug.cpp:738]: I think you should set `fpregs` to null if the original `fpregs` is invalid. The message in `print_context` will be confusing otherwise.
But I'll leave it to you assuming it'll never happen. Personally I think we should not try to avoid the assertion in `pointer_delta` and keep the code concise.

**Review by reinrich — APPROVED** [review]

File: src/hotspot/share/utilities/debug.cpp
Comment (TheRealMDoerr) [reply src/hotspot/share/utilities/debug.cpp:738]: I think null would be invalid, too. I prefer not touching it if it's already broken.

**Review by tstuefe — APPROVED** [review]
Okay!

**Comment by TheRealMDoerr** [comment]
Thanks again for the reviews! I didn't expect that printing a couple of register could be so difficult :-)
/integrate

**Pull request closed without merging**
